### DISCUSSION TITLE
Com qual começamos: Back ou Front

### DISCUSSION THREAD

**Discussion by eletroswing** [diego3g/rsxp-2023#2] [General]
## Olá pessoas!

Só queria esclarecer e saber por onde começaremos, com o backend nest/rest ou com o front-end (minha opinião seria o back, mas estou curioso e ansioso por isso)!

**Comment by jesse3g** [comment]
Acho que o back seria bom tbm.

**Reply by viniseven** [reply]
Se a comunidade se dividir pelo que tem mais afinidade digamos assim, daria para produzir os dois lados simultaneamente.

**Reply by jesse3g** [reply]
Verdade, séria ótimo isso.

**Reply by Gustavohsdp** [reply]
Justamente, mas acho que primeiro teria que decidir o schema do prisma com o Diego.

**Reply by eletroswing** [reply]
> Justamente, mas acho que primeiro teria que decidir o schema do prisma com o Diego.

Exatamente! Se não no final teremos incompatibilidades entre as duas partes, por isso estou curioso, um deve tocar o outro e nao ambos seguindo separado de inicio.

**Reply by FelipeQueiroz** [reply]
Com isso podemos até implementar mais features :)

**Comment by goncadanilo** [comment]
Acredito que conseguimos tocar os dois simultaneamente.

**Reply by eletroswing** [reply]
@viniseven @goncadanilo  os dois podem até andar juntos, mas por ser um rest, precisaria definir quais os endpoints a serem chamados e o que retornaria em cada um deles, afim de garantir a compatibilidade entre os dois!

**Reply by renaldodev** [reply]
Acho que se tivemos um roadmap do projecto com tarefas bem específicas e onde podemos ver o estado atual do projecto, qualquer um poderia contribuiria de forma mais aberta onde se sente mais a vontade de o fazer(back-end ou front-end).

**Reply by RennanD** [reply]
@eletroswing Acho que rodar os dois simultaneamente é a melhor opção, pois a gente não precisa necessariamente integrar o front com o back logo de cara.
Para simular no front podemos usar json-server ou algo do tipo e depois só ajustar os endpoint e respostas integrando com backend.

**Reply by brunoesposito** [reply]
o @eletroswing falou uma parada importante. Se for simultaneo ambos precisam saber o contrato de envio e resposta para nao precisar modificar nada no meio do caminho.

**Reply by brunoesposito** [reply]
@RennanD e esse json-server precisa ser definido e fixado logo no inicio. Para que o front nao force o back alterar algum contrato ou vice e versa.

**Reply by RennanD** [reply]
@brunoesposito Isso é uma boa, mas creio que não seja algo para travar o desenvolvimento do front, pois as telas em si não "dependem" do back.
Se formos esperar o backend ser finalizado e todo definido vamos, perder um tempo bom na criação do App. Se tivermos pelo menos as interfaces já desenvolvidas, com fluxo de navegação e essas coisas mais básicas desenvolvidas, vamos ganhar bastante tempo.

**Reply by nitoba** [reply]
Em meio a isso, o GraphQL seria uma opção melhor que padrão Rest por exemplo, pois o front end que iria pedir as informações que precisaria sem o backend ter que mudar a resposta das rotas, mas como estamos indo de Rest, é necessário ir fazendo esses ajustes ao longo do caminho enquanto está sendo desenvolvido. O lado bom é que ambos times de front e back estarão muito próximos.

**Comment by ikidoncc** [comment]
A melhor opção é realmente ser simultaneamente, já que o prazo é curso pra essa primeira entrega, além de que podemos ver cada deve mostrando seu real potencial na sua área, seja ela Back End ou Front End

**Comment by LuizHAP** [comment]
Tenho uma opinião melhor! Queria ver o que vocês acham:

Podemos criar os contratos entre backend/frontend, e podemos mockar utilizando algumas ferramentas de service worker ou de outro modo (https://mswjs.io/). E podemos paralelizar o desenvolvimento de ambos.

O que acham?

**Reply by brunoesposito** [reply]
Ferramente interessante!

**Reply by diego3g** [reply]
Acredito que o MSW não funcione pra desenvolvimento no RN por não suportar service workers, apenas pros testes que executam em um ambiente Node.js. Se formos seguir desenvolvendo com mocks, podemos pensar em usar o MirageJS ou alguma ferramenta semelhante.

**Comment by diego3g** [comment]
Podemos discutir isso na próxima live, mas acredito que, se montarmos bem direitinho os contratos, podemos desenvolver ambas simultâneamente.

**Reply by melquisedeque-magalhaes** [reply]
Acredito TMB Diego, que pode seguir o back e o front em paralelo, sem nem precisar de mock, e os dois vão evoluindo juntos, só organizar tudo certinho que quando finalizar os layout já vai ter rotas para consumir.

**Reply by leoferreiralima** [reply]
Entendo que podemos seguir com os dois em paralelo também, talvez uma ideia é criar um projeto separarado dentro do monorepo com os contratos e o zod conforme vimos na live, dessa maneira poderiamos importar  esse modulo dentro do back e front e algumas validações mais simples podereriam ser feitas mesmo antes de chamar a api, dessa maneira também garantiriamos que qualquer mudança nos contratos sejam refletidos no back e no front. Não sei se existe suporte no nest js para isso mas podemos pensar em usar o trpc para facilitar essa integração também

**Reply by diego3g** [reply]
Estou criando uma proposta de contratos usando o que mostrei na live hoje para trabalharmos a partir daí.
